### DISCUSSION TITLE
Including the 'correct' pico.h

### DISCUSSION THREAD

**Discussion by Jeff-Birt** [earlephilhower/arduino-pico#2627] [Q&A]
Just starting with Pico. Evaluating using VS Code or Arduino IDE. Tried to copy LED blink example VS Code to Arduino. Got hung up on the 'correct' pico.h not being included. In the VS Code extension config_autogen.h has this line

#include "<user_path>/.pico-sdk/sdk/2.0.0/src/boards/include/boards/pico.h"

I find nothing similar in the Arduino IDE and cannot at this juncture deduce how to have this 'correct' pico.h included for each RP2040 based project.

Help appreciated.

**Comment by maxgerhardt** [comment]
If you just install the Arduino-Pico core in the Arduino IDE [as documented](https://github.com/earlephilhower/arduino-pico?tab=readme-ov-file#installation), then a simple [Blink sketch](https://github.com/arduino/arduino-examples/blob/main/examples/01.Basics/Blink/Blink.ino) should not come up with an error about not finding `pico.h`. This points to you having a corrupted version of this core.

How exactly did you install this core, how do you trigger a build?

**Reply by Jeff-Birt** [reply]
Installed via board manage per https://github.com/earlephilhower/arduino-pico

Blink example IS NOT the one from Arduino IDE, as mentioned in my post. The problem is that the definition for 'PICO_DEFAULT_LED_PIN' is not found. Below is the init function of the blink example. I built with the build/verify button like any other Arduino sketch. 

As mentioned in my original post the correct pico.h is not being included for the board (RP2040).

// Perform initialisation
int pico_led_init(void) {
#if defined(PICO_DEFAULT_LED_PIN)
    // A device like Pico that uses a GPIO for the LED will define PICO_DEFAULT_LED_PIN
    // so we can use normal GPIO functionality to turn the led on and off
    gpio_init(PICO_DEFAULT_LED_PIN);
    gpio_set_dir(PICO_DEFAULT_LED_PIN, GPIO_OUT);
    return PICO_OK;
#elif defined(CYW43_WL_GPIO_LED_PIN)
    // For Pico W devices we need to initialise the driver etc
    return cyw43_arch_init();
#endif
}

**Reply by maxgerhardt** [reply]
..No, you can't just copy paste a Pico-SDK example verbatim into the Arduino IDE and this Arduino core and expect it to work. See https://arduino-pico.readthedocs.io/en/latest/sdk.html. 

This Arduino core only includes the Pico-SDK as a precompiled library. Specifically during copilation for the RP2040, which is used for **all** RP2040 boards, it selects `PICO_BOARD pico_w` (see `tools/libpico/CMakeLists.txt`). The core then has different variants that contain a `pins_arduino.h` file, which declares all the pins for the board, including the LED if present.

The core then does not include the logic to auto-include the right board.h file associated with the board. All pins definition **it** needs are in the `pins_arduino.h`. As a trick with the include path, you can get the pin definitions for a Pico with
```cpp
#include <../../../boards/include/boards/pico.h>
```

So, the "correct" pico.h, or rather, the "correct" boards file essentially **is** `pins_arduino.h`. It will of course create other macros and variables then `boards/pico.h`, you will have to adapt the code for it.

https://github.com/earlephilhower/arduino-pico/blob/996c3bfab9758f12c07aa20cc6d352e630c16987/variants/rpipico/pins_arduino.h#L7-L15

You should really consider using **Arduino** code with this Arduino core, and not pure Pico-SDK code; As linked in the documentation, calling Pico-SDK functions can badly mess with the Arduino core if intermixed.

**Reply by maxgerhardt** [reply]
After all, writing
```cpp
#include "pico/stdlib.h"
#include <../../../boards/include/boards/pico.h> // hardcoded for: Pico board
                                               // use pins_arduino.h and adapt as needed
#ifdef CYW43_WL_GPIO_LED_PIN
#include "pico/cyw43_arch.h"
#endif

#ifndef LED_DELAY_MS
#define LED_DELAY_MS 250
#endif

int pico_led_init(void) {
#if defined(PICO_DEFAULT_LED_PIN)
  // A device like Pico that uses a GPIO for the LED will define PICO_DEFAULT_LED_PIN
  // so we can use normal GPIO functionality to turn the led on and off
  gpio_init(PICO_DEFAULT_LED_PIN);
  gpio_set_dir(PICO_DEFAULT_LED_PIN, GPIO_OUT);
  return PICO_OK;
#elif defined(CYW43_WL_GPIO_LED_PIN)
  // For Pico W devices we need to initialise the driver etc
  return cyw43_arch_init();
#endif
}


// Turn the led on or off
void pico_set_led(bool led_on) {
#if defined(PICO_DEFAULT_LED_PIN)
  // Just set the GPIO on or off
  gpio_put(PICO_DEFAULT_LED_PIN, led_on);
#elif defined(CYW43_WL_GPIO_LED_PIN)
  // Ask the wifi "driver" to set the GPIO on or off
  cyw43_arch_gpio_put(CYW43_WL_GPIO_LED_PIN, led_on);
#endif
}

void setup() {
  pico_led_init();
}

void loop() {
  pico_set_led(true);
  sleep_ms(LED_DELAY_MS);
  pico_set_led(false);
  sleep_ms(LED_DELAY_MS);
}
```
is so much less readable and unnecessarily complicated than the pure and simple

```cpp
void setup() {
  // initialize digital pin LED_BUILTIN as an output.
  pinMode(LED_BUILTIN, OUTPUT);
}

// the loop function runs over and over again forever
void loop() {
  digitalWrite(LED_BUILTIN, HIGH);  // turn the LED on (HIGH is the voltage level)
  delay(1000);                      // wait for a second
  digitalWrite(LED_BUILTIN, LOW);   // turn the LED off by making the voltage LOW
  delay(1000);                      // wait for a second
}
```

Which automatically works correctly for a Pico and Pico W, no weird compile time checks needed in the sketch.

**Reply by Jeff-Birt** [reply]
Thanks for the information. The point of my simple exercise was to see what it would take to port some bit of code from Pico SDK to Arduino Pico core. The Arduinofication of defines or standard pins/names is not a showstopper. Found interesting bug when debugging but will start a new discussion for that.
